### ISSUE TITLE
Using null in AWS::Serverless::StateMachine in the ResultPath throws incorrect E0000 error

### ISSUE THREAD

**Issue by dontirun** [aws-cloudformation/cfn-lint#1657]
StateMachines allow for a null value in the ResultPath Property to [pass input directly to the output](https://docs.aws.amazon.com/step-functions/latest/dg/input-output-resultpath.html)  

Below is a minimal example of a template that is proper but flagged as invalid

```
AWSTemplateFormatVersion: 2010-09-09
Transform: AWS::Serverless-2016-10-31

Resources:
  rLambda:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.handler
      Runtime: python3.6
      InlineCode: |
        def handler(event, context):
          print(event)
          return


  rTestMachine:
    Type: AWS::Serverless::StateMachine
    Properties:
      Definition: 
        StartAt: myState
        States:
          myState:
            Type: Task
            Resource:  !GetAtt rLambda.Arn
            ResultPath: null
            End: true
      Policies:
      - Statement:
        - Sid: Test
          Effect: Allow
          Action:
          - lambda:Invoke*
          Resource: !GetAtt rLambda.Arn
```

Adding quotes around null does not resolve the issue as the value of the property needs to evaluate to null. An exception for null checking should be added for this Parameter

**Comment by kddejong** [comment]
Thanks for submitting this issue.

For my tracking this is the translated template.  This gets translated to string which masks the null to the CloudFormation service.  The cfn-lint ordering may be adjust to handle this situation correctly.  
```
   "DefinitionString": {
          "Fn::Join": [
            "\n",
            [
              "{",
              "    \"StartAt\": \"myState\",",
              "    \"States\": {",
              "        \"myState\": {",
              "            \"End\": true,",
              "            \"Resource\": \"${definition_substitution_1}\",",
              "            \"ResultPath\": null,",
              "            \"Type\": \"Task\"",
              "        }",
              "    }",
              "}"
            ]
          ]
        },
```

**Comment by kddejong** [comment]
If this remained as yaml or json (not as a string) and wasn't translated by SAM there would be a `[/Resources/rTestMachine/Type/DefinitionString/States/myState/ResultPath] 'null' values are not allowed in templates` error.

**Comment by sc-alscient** [comment]
Hi is there any workaround for this until the bug is resolved? Other than using json! Thanks.

**Comment by PatMyron** [comment]
Unfortunately, the [Serverless Transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) drops [attributes like `Metadata`](https://github.com/aws/serverless-application-model/issues/450#issuecomment-643420308) for [`AWS::Serverless` resource types](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resources-and-properties.html)

**Comment by sc-alscient** [comment]
Thanks for getting back to me so quickly. Does this apply for whole template as well? I tried the below but still have the same problem:
```
AWSTemplateFormatVersion: "2010-09-09"
Transform: AWS::Serverless-2016-10-31
Description: UI Cypress testing resources

Metadata:
  cfn-lint:
    config:
      ignore_checks:
        - E0000
 ```

**Comment by mndeveci** [comment]
Hi,

With [SAM v1.36.0](https://github.com/aws/serverless-application-model/releases/tag/v1.36.0) and [SAM CLI v1.24.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.24.0) versions, we now support adding Metadata to the serverless resources.

Can you verify if that will help with your use case?

Thanks!

**Comment by edmundcong** [comment]
Any ETA as to when this will be fixed? it's forcing our step after our map to handle our incoming input in a remarkably awkward way

**Comment by tianmarin** [comment]
Maybe i'm missing something, this issue is closed because there is no error thrown now, but yaml template with `ResultPath: null` won't create a state with a `ResultPath` equals to `null`, it will only omit the `ResultPath` parameter.

**Comment by kddejong** [comment]
This is the template returned by running `cfn-lint template.yaml --info`

```json
{
  "rTestMachine": {
    "Properties": {
      "DefinitionString": {
       "Fn::Join": [
        "\n",
        [
         "{",
         "    \"StartAt\": \"myState\",",
         "    \"States\": {",
         "        \"myState\": {",
         "            \"End\": true,",
         "            \"Resource\": \"${definition_substitution_1}\",",
         "            \"ResultPath\": null,",
         "            \"Type\": \"Task\"",
         "        }",
         "    }",
         "}"
        ]
       ]
      },
      "DefinitionSubstitutions": {
       "definition_substitution_1": {
        "Fn::GetAtt": [
         "rLambda",
         "Arn"
        ]
       }
      },
      "RoleArn": {
       "Fn::GetAtt": [
        "rTestMachineRole",
        "Arn"
       ]
      },
      "Tags": [
       {
        "Key": "stateMachine:createdBy",
        "Value": "SAM"
       }
      ]
     },
     "Type": "AWS::StepFunctions::StateMachine"
   }
}
```

**Comment by tianmarin** [comment]
Thank tou very much for your quick response!


Then it's a CloudFormation error?


On CF console when I see the template it is shown like this:
```yaml
                SaveInfoToBucket:
                  Type: Task
                  ResultPath: null
                  Next: OtherStep
                  Retry:
                  - ErrorEquals:
                    - S3.SdkClientException
                    IntervalSeconds: 3
                    MaxAttempts: 3
                  Parameters:
                    Body.$: $
                    Bucket:
                      Ref: BucketId
                    Key.$: States.Format('{}_{}.json', $$.Execution.StartTime, $.field)
                  Resource: arn:aws:states:::aws-sdk:s3:putObject
```
But if i click **View processed template**:
```
              "                   \"SaveInfoToBucket\": {",
              "                        \"Next\": \"OtherStep\",",
              "                        \"Parameters\": {",
              "                            \"Body.$\": \"$\",",
              "                            \"Bucket\": \"${definition_substitution_12}\",",
              "                            \"Key.$\": \"States.Format('{}_{}.json', $$.Execution.StartTime, $.field)\"",
              "                        },",
              "                        \"Resource\": \"arn:aws:states:::aws-sdk:s3:putObject\",",
              "                        \"Retry\": [",
              "                            {",
              "                                \"ErrorEquals\": [",
              "                                    \"S3.SdkClientException\"",
              "                                ],",
              "                                \"IntervalSeconds\": 3,",
              "                                \"MaxAttempts\": 3",
              "                            }",
              "                        ],",
              "                        \"Type\": \"Task\"",
              "           
```

**Comment by kddejong** [comment]
That is kind of baffling to me as the sam translator library from pypi will convert it to this:
```json
"DefinitionString": {
     "Fn::Join": [
      "\n",
      [
       "{",
       "    \"StartAt\": \"myState\",",
       "    \"States\": {",
       "        \"SaveInfoToBucket\": {",
       "            \"Next\": \"OtherStep\",",
       "            \"Parameters\": {",
       "                \"Body.$\": \"$\",",
       "                \"Bucket\": \"${definition_substitution_1}\",",
       "                \"Key.$\": \"States.Format('{}_{}.json', $$.Execution.StartTime, $.field)\"",
       "            },",
       "            \"Resource\": \"arn:aws:states:::aws-sdk:s3:putObject\",",
       "            \"ResultPath\": null,",
       "            \"Retry\": [",
       "                {",
       "                    \"ErrorEquals\": [",
       "                        \"S3.SdkClientException\"",
       "                    ],",
       "                    \"IntervalSeconds\": 3,",
       "                    \"MaxAttempts\": 3",
       "                }",
       "            ],",
       "            \"Type\": \"Task\"",
       "        }",
       "    }",
       "}"
      ]
     ]
    },
```

They don't look familiar this may be an issue unrelated to cfn-lint since we rely on SAM to transform the template appropriately.

**Comment by tianmarin** [comment]
@kddejong Thank you very much again for your quick response

After your response, I checked every tiny thing that might be causing this and voilà 🥳 

My template uses two transformations:
```yaml
---
AWSTemplateFormatVersion: 2010-09-09
Transform: 
  - 'AWS::LanguageExtensions'
  - 'AWS::Serverless-2016-10-31'
```
They are in the correct order as stated here:
[Introducing new language extensions in AWS CloudFormation](https://aws.amazon.com/es/blogs/mt/introducing-new-language-extensions-in-aws-cloudformation/)

Once I removed the `AWS::LanguageExtensions` (as it is not used in my `AWS::Serverless::StateMachine`resources), the transformation of SAM works perfectly:
```
              "     },",
              "     \"Resource\": \"arn:aws:states:::aws-sdk:s3:putObject\",",
              "     \"ResultPath\": null,",
              "     \"Retry\": [",
              "     {",
```


I hope this is useful for someone else.

**Comment by kddejong** [comment]
that is useful and I wonder if its a bug in how the `AWS::LanguageExtensions` works.

**Comment by niokolra** [comment]
I now face the same issue when the cf is in YAML and the definition is JSON. I am unable to deploy my step function while ResultPath value is null

**Comment by dontirun** [comment]
@niokolra I don't think the deployment error is related to `cfn-lint`. 

CloudFormation will throw an error if it detects `null` values in YAML. You can get around that by switching from `AWS::StepFunctions::StateMachine` to `AWS::Serverless::StateMachine` and using the [serverless transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) in your CloudFormation template. 

This gets around the issue  because behind the scenes the Serverless Transform converts the  YAML specified in the`AWS::Serverless::StateMachine`  to stringified JSON before attempting to deploy the template

**Comment by dontirun** [comment]
I misread the question, ignore my previous comment  🤦

**Comment by kddejong** [comment]
@niokolra can you provide a few more details.  Are you using the language transform?  Are you using SAM?  Or is it just a CloudFormation template with a step function defined in JSON

**Comment by niokolra** [comment]
@kddejong this is just a CloudFormation template with a step function defined in JSON. When I try to deploy it I receive the error

**Comment by kddejong** [comment]
@niokolra do you get that error from the service or cfn-lint?

**Comment by MalikAtalla-AWS** [comment]
Just for visibility: [The underlying issue](https://github.com/aws-cloudformation/cfn-language-discussion/issues/100)  in the `AWS::LanguageExtensions` Transform has been fixed.

**Comment by mbwilding** [comment]
> I misread the question, ignore my previous comment 🤦

Hey, you solved it for me, I swapped to the serverless version, thanks for that.
